src/charm.py

```diff
@@ -0,0 +1,301 @@
+#!/usr/bin/env python3
+# Copyright 2021 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+import glob
+import time
+from typing import Optional
+
+import logging
+from pathlib import Path
+
+from ops.charm import CharmBase
+from ops.main import main
+from ops.model import ActiveStatus, WaitingStatus, MaintenanceStatus, BlockedStatus
+
+import lightkube
+from lightkube import codecs
+from lightkube.resources.apps_v1 import StatefulSet
+from lightkube.core.exceptions import ApiError
+
+
+METACONTROLLER_IMAGE = "metacontroller/metacontroller:v0.3.0"
+
+
+class MetacontrollerOperatorCharm(CharmBase):
+    """Charm the Metacontroller"""
+
+    def __init__(self, *args):
+        super().__init__(*args)
+
+        if not self.unit.is_leader():
+            self.model.unit.status = WaitingStatus("Waiting for leadership")
+            return
+
+        self.framework.observe(self.on.install, self._install)
+        self.framework.observe(self.on.remove, self._remove)
+        self.framework.observe(self.on.update_status, self._update_status)
+        # self.framework.observe(self.on.config_changed, self._reconcile)
+
+        self.logger = logging.getLogger(__name__)
+
+        # TODO: Fix file imports and move ./src/files back to ./files
+        self._manifests_file_root = None
+        self.manifest_file_root = "./src/files/manifests/"
+
+        self._lightkube_client = None
+
+    def _install(self, event):
+        self.logger.info("Installing by instantiating Kubernetes objects")
+        self.unit.status = MaintenanceStatus("Instantiating Kubernetes objects")
+
+        # TODO: catch error when this fails due to permissions and set appropriate "we aren't
+        #  trusted" blocked status
+        # create rbac
+        try:
+            self._create_rbac()
+        except ApiError as e:
+            # Handle forbidden error as this likely means we don't have --trust
+            if e.status.code == 403:
+                self.logger.error(
+                    "Received Forbidden (403) error from lightkube when creating required RBAC.  "
+                    "This may be due to the charm lacking permissions to create cluster-scoped "
+                    "roles and resources.  Charm must be deployed with `--trust`"
+                )
+                self.unit.status = BlockedStatus(
+                    "Cannot create required RBAC.  Charm may not have `--trust`"
+                )
+                return
+            else:
+                raise e
+
+        # Create crds
+        self._create_crds()
+
+        # deploy the controller
+        self._create_controller()
+
+        self.logger.info("Waiting for installed Kubernetes objects to be operational")
+        max_attempts = 20
```

Comment:
I was thinking the same.  Will put on the todo list for Monday

Reply:
I agree with tenacity being a good idea, but if this takes up too much time lets stick with the current solution and revisit afterwards.

Reply:
Didn't think it would take much time, but then it did... :D  But it is there now.

It doesn't make the code any easier to read though imo.  `retry` gets a bit awkward and verbose to me